core/clustersmngr/factory.go

```diff
@@ -0,0 +1,306 @@
+	UpdateNamespaces(ctx context.Context) error
+	// GetServerClient returns the cluster client with gitops server permissions
+	GetServerClient(ctx context.Context) (Client, error)
+	// GetClustersNamespaces returns the namespaces for all clusters
+	GetClustersNamespaces() map[string][]v1.Namespace
+	// Start starts go routines to keep clusters and namespaces lists up to date
+	Start(ctx context.Context)
+}
+
+type clientsFactory struct {
+	hubClient       client.Client
+	clustersFetcher ClusterFetcher
+	nsChecker       nsaccess.Checker
+	log             logr.Logger
+
+	// list of clusters returned by the clusters fetcher
+	clusters *Clusters
+	// the lists of all namespaces of each cluster
+	clustersNamespaces *ClustersNamespaces
+	// lists of namespaces accessible by the user on every cluster
+	usersNamespaces *UsersNamespaces
+}
+
+func NewClientFactory(hubClient client.Client, fetcher ClusterFetcher, nsChecker nsaccess.Checker, logger logr.Logger) ClientsFactory {
+	return &clientsFactory{
+		hubClient:          hubClient,
+		clustersFetcher:    fetcher,
+		nsChecker:          nsChecker,
+		clusters:           &Clusters{},
+		clustersNamespaces: &ClustersNamespaces{},
+		usersNamespaces:    &UsersNamespaces{Cache: ttlcache.New(24 * time.Hour)},
+		log:                logger,
+	}
+}
+
+func (cf *clientsFactory) Start(ctx context.Context) {
+	go cf.watchClusters(ctx)
+	go cf.watchNamespaces(ctx)
+}
+
+func (cf *clientsFactory) watchClusters(ctx context.Context) {
+	if err := wait.PollImmediateInfinite(watchClustersFrequency, func() (bool, error) {
+		if err := cf.UpdateClusters(ctx); err != nil {
+			return false, err
+		}
+
+		return true, nil
+	}); err != nil {
+		cf.log.Error(err, "failed pooling clusters")
+	}
+}
+
+func (cf *clientsFactory) UpdateClusters(ctx context.Context) error {
+	clusters, err := cf.clustersFetcher.Fetch(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to fetch clusters: %w", err)
+	}
+
+	cf.clusters.Set(clusters)
+
+	return nil
+}
+
+func (cf *clientsFactory) watchNamespaces(ctx context.Context) {
+	if err := wait.PollImmediateInfinite(watchNamespaceFrequency, func() (bool, error) {
+		if err := cf.UpdateNamespaces(ctx); err != nil {
+			return false, err
+		}
+
+		return false, nil
+	}); err != nil {
+		cf.log.Error(err, "failed pooling namespaces")
+	}
+}
+
+func (cf *clientsFactory) UpdateNamespaces(ctx context.Context) error {
+	clients, err := clientsForClusters(cf.clusters.Get())
+	if err != nil {
+		cf.log.Error(err, "failed to create clients for", "clusters", cf.clusters.Get())
+		return err
+	}
+
+	wg := sync.WaitGroup{}
+
+	for clusterName, c := range clients {
+		wg.Add(1)
+
+		go func(clusterName string, c client.Client) {
+			defer wg.Done()
+
+			nsList := &v1.NamespaceList{}
+
+			if err := c.List(ctx, nsList); err != nil {
+				cf.log.Error(err, "failed listing namespaces", "cluster", clusterName)
+			}
+
+			cf.clustersNamespaces.Set(clusterName, nsList.Items)
```

Comment:
do we want to leave in the case of error? what would happen in the case that `nsList.Items` is `0` (can it be `nil`?) and it is added to the map?

Reply:
no problem it will just be set as an empty list and web move on since we need to get the other namespaces

Reply:
sweet 👍